sdk/core/azure-core-amqp/vendor/azure-uamqp-c/src/connection.c

```diff
@@ -173,49 +173,64 @@ static int channel_table_allocate(CHANNEL_TABLE* channel_table, uint16_t* h)
+                result = MU_FAILURE;
+            }
+            else
+            {
+                uint32_t new_capacity = channel_table->capacity + INITIAL_CHANNEL_TABLE_CAPACITY;
+                if (channel_table->capacity >= UINT16_MAX)
+                {
+                    new_capacity = UINT16_MAX;
+                }
+                CHANNEL_TABLE_ENTRY* new_channels = (CHANNEL_TABLE_ENTRY*)realloc(
+                    channel_table->channels, sizeof(CHANNEL_TABLE_ENTRY) * new_capacity);
+                if (new_channels == NULL)
+                {
+                    LogError("Cannot reallocate memory for handle table");
+                    result = MU_FAILURE;
+                }
+                else
+                {
+                    channel_table->channels = new_channels;
+                    channel_table->capacity = new_capacity;
+                    // Ensure the newly resized part of the handle table is empty.
+                    uint32_t i = channel_table->size;
+                    do
+                    {
+                        channel_table->channels[i].outgoing_channel = ENDPOINT_UNUSED_CHANNEL;
+                        channel_table->channels[i].incoming_channel = ENDPOINT_UNUSED_CHANNEL;
+                        channel_table->channels[i].is_endpoint_live = false;
+                        i++;
+                    } while (i < channel_table->capacity);
+                }
             }
-            channel_table->capacity *= 2;
-          }
         }
         if (!result)
         {
-          uint16_t i = 0;
-          for (; i < channel_table->size; i++)
-          {
-            if (channel_table->channels[i].outgoing_channel == ENDPOINT_UNUSED_CHANNEL)
+            // Look for an empty slot in the table. If none is found, add to the end of the table.
+            uint32_t i = 0;
+            for (; i < channel_table->size; i++)
+            {
+                if (channel_table->channels[i].outgoing_channel == ENDPOINT_UNUSED_CHANNEL)
+                {
+                    break;
+                }
+            }
+            // If we didn't find a hole, we need to increase the size of the table by 1.
+            if (i == channel_table->size)
             {
-              *h = i;
-              channel_table->channels[i].outgoing_channel = i;
-              break;
+                channel_table->size++;
             }
-          }
-          // If we didn't find a hole, add this to the end of the table.
-          if (i == channel_table->size)
-          {
-            *h = channel_table->size;
-            channel_table->channels[channel_table->size].outgoing_channel = channel_table->size;
-            channel_table->channels[channel_table->size].is_endpoint_live = true;
-            channel_table->size++;
-          }
+
+            *h = (uint16_t)i;
+            channel_table->channels[i].outgoing_channel = (uint16_t)i;
+            channel_table->channels[i].is_endpoint_live = true;
```

Comment:
The question of some kind, `[i]`, or `[channel_table->size]`?

Reply:
(and `channel_table->channels[i].outgoing_channel = (uint16_t)i;` and not `channel_table->channels[channel_table->size].outgoing_channel = (uint16_t)channel_table->size;`)

Reply:
`i` is the index into the table where we found an unused channel.
`channel_table->size` is highest valid entry in the table.

Arguably there's no point in having `channel_table->size`, since it goes up but never down.

Reply:
I see now. Thanks!